pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1015 @@
+	for peer := range peerKeys {
+		peerConfigs = append(peerConfigs, allPeerConfigs[peer])
+	}
+	return peerConfigs
+}
+
+func (c *Controller) reconcileBGPPeers(curPeerKeys sets.Set[string],
+	curPeerConfigs map[string]bgp.PeerConfig,
+	bpState *bgpPolicyState,
+	bgpServerUpdated bool) error {
+
+	prePeerKeys := bpState.peerKeys
+	prePeerConfigs := bpState.peerConfigs
+
+	var peerToAddKeys sets.Set[string]
+	if !bgpServerUpdated {
+		peerToAddKeys = curPeerKeys.Difference(prePeerKeys)
+	} else {
+		peerToAddKeys = curPeerKeys
+	}
+	peerConfigsToAdd := getPeerConfigs(peerToAddKeys, curPeerConfigs)
+	for _, peer := range peerConfigsToAdd {
+		if err := bpState.bgpServer.AddPeer(c.ctx, peer); err != nil {
+			return err
+		}
+	}
+
+	if !bgpServerUpdated {
+		peerToUpdateKeys := sets.New[string]()
+		remainPeerKeys := prePeerKeys.Intersection(curPeerKeys)
+		for peerKey := range remainPeerKeys {
+			prevPeerConfig := bpState.peerConfigs[peerKey]
+			curPeerConfig := curPeerConfigs[peerKey]
+			if !reflect.DeepEqual(prevPeerConfig, curPeerConfig) {
+				peerToUpdateKeys.Insert(peerKey)
+			}
+		}
+		peerToUpdateConfigs := getPeerConfigs(peerToUpdateKeys, curPeerConfigs)
+		for _, peer := range peerToUpdateConfigs {
+			if err := bpState.bgpServer.UpdatePeer(c.ctx, peer); err != nil {
+				return err
+			}
+		}
+
+		peerToDeleteKeys := prePeerKeys.Difference(curPeerKeys)
+		peerToDeleteConfigs := getPeerConfigs(peerToDeleteKeys, prePeerConfigs)
+		for _, peer := range peerToDeleteConfigs {
+			if err := bpState.bgpServer.RemovePeer(c.ctx, peer); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *Controller) reconcileRoutes(curRoutes sets.Set[bgp.Route], bpState *bgpPolicyState, bgpServerUpdated bool) error {
+	prevRoutes := bpState.routes
+
+	var routesToAdvertise sets.Set[bgp.Route]
+	if !bgpServerUpdated {
+		routesToAdvertise = curRoutes.Difference(prevRoutes)
+	} else {
+		routesToAdvertise = curRoutes
+	}
+	if routesToAdvertise.Len() != 0 {
+		if err := bpState.bgpServer.AdvertiseRoutes(c.ctx, routesToAdvertise.UnsortedList()); err != nil {
+			return err
+		}
+	}
+
+	if !bgpServerUpdated {
+		routesToWithdraw := prevRoutes.Difference(curRoutes)
+		if routesToWithdraw.Len() != 0 {
+			if err := bpState.bgpServer.WithdrawRoutes(c.ctx, routesToWithdraw.UnsortedList()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) getBGPPolicyState(bpName string) (*bgpPolicyState, bool) {
+	c.bgpPolicyStatesMutex.RLock()
+	defer c.bgpPolicyStatesMutex.RUnlock()
+	state, exists := c.bgpPolicyStates[bpName]
+	return state, exists
+}
+
+func (c *Controller) deleteBGPPolicyState(bpName string) {
+	c.bgpPolicyStatesMutex.Lock()
+	defer c.bgpPolicyStatesMutex.Unlock()
+	delete(c.bgpPolicyStates, bpName)
+}
+
+func (c *Controller) getRouterID() (string, error) {
+	var routerID string
+	// For IPv6 only environment, the BGP routerID should be specified by K8s Node annotation `antrea. io/ bgp-route-id`.
```

Comment:
```suggestion
	// For IPv6 only environment, the BGP routerID should be specified by K8s Node annotation `antrea.io/bgp-route-id`.
```
Question: if this is provided by users, how to avoid duplicate IDs?

Reply:
We cannot avoid duplicate IDs.

Reply:
Will there be any side-affect if there are duplicate IDs?